inmem/kv.go

```diff
@@ -232,3 +239,143 @@ func (b *Bucket) getAll(o *kv.CursorHints) ([]kv.Pair, error) {
 
 	return pairs, nil
 }
+
+type pair struct {
+	kv.Pair
+	err error
+}
+
+// ForwardCursor returns a directional cursor which starts at the provided seeked key
+func (b *Bucket) ForwardCursor(seek []byte, opts ...kv.CursorOption) (kv.ForwardCursor, error) {
+	var (
+		pairs = make(chan []pair)
+		stop  = make(chan struct{})
+	)
+
+	go func() {
+		defer close(pairs)
+
+		var (
+			config  = kv.NewCursorConfig(opts...)
+			fn      = config.Hints.PredicateFn
+			iterate = func(it btree.ItemIterator) {
+				b.btree.AscendGreaterOrEqual(&item{key: seek}, it)
+			}
+		)
+
+		if config.Direction == kv.CursorDescending {
+			iterate = func(it btree.ItemIterator) {
+				b.btree.DescendLessOrEqual(&item{key: seek}, it)
+			}
+		}
+
+		var batch []pair
+
+		iterate(func(i btree.Item) bool {
+			select {
+			case <-stop:
+				// if signalled to stop then exit iteration
+				return false
+			default:
+			}
+
+			j, ok := i.(*item)
+			if !ok {
+				batch = append(batch, pair{err: fmt.Errorf("error item is type %T not *item", i)})
+
+				return false
+			}
+
+			if fn == nil || fn(j.key, j.value) {
+				batch = append(batch, pair{Pair: kv.Pair{Key: j.key, Value: j.value}})
+			}
+
+			if len(batch) < cursorBatchSize {
+				return true
+			}
+
+			pairs <- batch
```

Comment:
@stuartcarnie I actually do this at the top of the function. I could move it down.
I would be worried if I moved it down though that it would mean it tries to fill the batch before acknowledging the stop signal. Since this comes after a length check on the batch as this function is called once per item in the interation.

I guess the best situation would be to do this everywhere. Once on enter incase since the last item we were signalled to stop. Once on send to avoid sending a batch on a close. Once on the final flush.

Reply:
@GeorgeMac sorry if I was unclear, I didn't intend for you to remove it from the top. It sounds like you concluded the same thing, which is to probe it in a few places 👍 

This is similar what we see by periodically polling the `context.Done()` channel to determine if the context has been canceled in anyway.